packages/qwik/src/core/util/string.unit.ts

```diff
@@ -0,0 +1,46 @@
+import { test, expect } from 'vitest';
+import { packUint8Array, unpackUint8Array } from './string';
+import crypto from 'node:crypto';
+
+test('pack/unpack bytes including surrogate pair', () => {
+  const a = new Uint8Array([66, 216, 183, 223, 206, 145, 182, 91]);
+  const packed = packUint8Array(a);
+  expect(packed).toBe('𠮷野家');
+  expect(typeof packed).toBe('string');
+  expect(unpackUint8Array(packed)).toStrictEqual(a);
+});
+
+test('random pack/unpack of even array', () => {
+  const a = new Uint8Array(8);
+  const enc = new TextEncoder();
+  const dec = new TextDecoder();
+  for (let i = 0; i < 1000; i++) {
+    crypto.getRandomValues(a);
+    const packed = packUint8Array(a);
+    const code = enc.encode(packed);
+    const decoded = dec.decode(code);
+    expect(unpackUint8Array(decoded)).toStrictEqual(a);
+  }
+});
+
+test('random pack/unpack of odd array', () => {
+  const a = new Uint8Array(7);
+  const enc = new TextEncoder();
+  const dec = new TextDecoder();
+  for (let i = 0; i < 1000; i++) {
+    crypto.getRandomValues(a);
+    const packed = packUint8Array(a);
+    const code = enc.encode(packed);
+    const decoded = dec.decode(code);
+    expect(unpackUint8Array(decoded)).toStrictEqual(a);
+  }
+});
+
+test('space efficient', () => {
+  const a = new Uint8Array(65536);
+  crypto.getRandomValues(a);
+  const packed = packUint8Array(a);
+  // 0xD800-0xDFFF = 2048. 2048/65536 = 0.03125
+  // These are doubled in length, so inflating ratio is about 1.0625
+  expect((packed.length * 2) / a.length).toBeLessThan(1.07);
```

Comment:
It depends on the encoding of the HTML document.
If it uses UTF-16, our encoder/decoder reaches the best inflation ratio 106%.
But it uses UTF-8, it drops to about 147%.

in UTF-8:
* 1 byte: 0x00-0x7F: IR = 50%, share is 256/65536
* 2 bytes: 0x80-0x7FF:  IR = 100%, share is 1920/65536
* 3 bytes: 0x800-0xFFFF: IR = 150%, share is (63488 - 2048)/65536 = 61440/65536
* 4 bytes: 0x10000-0x10FFFF (all surrogate pairs in UTF-16) IR = 100%, 2048/65536
So totally,  IR is about 147%

If we can omit the support other than the UTF-8, the base64 may be better choice because of the UTF-8 is latin1 friendly encoding. The inflation ratio (IR) of the base64 encoding is 4/3, it's about 133%.
It may be the best way to choose the encoder/decoder depending on the HTML document.

The UTF-8 is the standard today, but if you think about efficiency on serializing random bytes, choosing UTF-16 is better.

Reply:
It depends on the encoding of the HTML document.
If it uses UTF-16, our encoder/decoder reaches the best inflation ratio (IR) 106%.
But it uses UTF-8, it drops to about 147%.

in UTF-8:
1 byte: 0x00-0x7F: IR = 50%, share is 256/65536
2 bytes: 0x80-0x7FF: IR = 100%, share is 1920/65536
3 bytes: 0x800-0xFFFF: IR = 150%, share is (63488 - 2048)/65536 = 61440/65536
4 bytes: 0x10000-0x10FFFF (all surrogate pairs in UTF-16) IR = 100%, 2048/65536
So totally, IR is about 147%

If Qwik omits the support other than the UTF-8, the base64 may be better choice because of the UTF-8 is latin1 friendly encoding. The IR of the base64 encoding is 4/3, it's about 133%.
It may be the best way to choose the encoder/decoder depending on the encoding of HTML document, for example, if it is UTF-8 or latin1, uses base64, otherwize this serializer.

Reply:
This result comes from the UTF-8 is not space efficient other than latin1 characters.
In spite of the UTF-8 is common for HTML, using UTF-16 is better choice if the document contains many characters over the 0x7f code points because most of 3 bytes in UTF-8 are 2 bytes in UTF-16.

Reply:
In particular the Qwik, as of using many serialization that produces special characters, it may be good to set the UTF-16 as default encoding for better performance.

Reply:
I think Qwik only supports utf-8 at the moment, so then base64 would be a better choice :-/.

However, we have 96 safe-ish characters at our disposal that encode to one byte, so we could implement our own base-96. However, [here's a very cool package](https://github.com/qntm/base65536?tab=readme-ov-file#:~:text=Note%20that%20in%20UTF%2D16%2C%20Base32768%20significantly%20outperforms%20either%20choice%2C%20and%20in%20UTF%2D8%2C%20Base64%20remains%20the%20preferred%20choice) that does UTF-32 encoding and it talks about other encodings, saying that for UTF-8, base-64 is still preferred :-/.